src/OSPSuite.Infrastructure/Container/Autofac/AutofacContainer.cs

```diff
@@ -0,0 +1,158 @@
+      {
+         if (serviceType.IsGenericTypeDefinition && concreteType.IsGenericTypeDefinition)
+            RegisterOpenGeneric(serviceType, concreteType);
+         else
+            Register(serviceType, concreteType, LifeStyle.Transient);
+      }
+
+      public void Register(Type serviceType, Type concreteType, LifeStyle lifeStyle)
+      {
+         Register(new[] {serviceType}, concreteType, lifeStyle);
+      }
+
+      public void Register(Type serviceType, Type concreteType, LifeStyle lifeStyle, string key)
+      {
+         register(new[] {serviceType}, concreteType, lifeStyle, key);
+      }
+
+      public IEnumerable<TInterface> ResolveAll<TInterface>() => Container.Resolve<IEnumerable<TInterface>>();
+
+      public void Register<TService1, TConcreteType>(LifeStyle lifeStyle)
+      {
+         Register(new[] {typeof(TService1)}, typeof(TConcreteType), lifeStyle);
+      }
+
+      public void Register<TService1, TService2, TConcreteType>(LifeStyle lifeStyle)
+      {
+         Register(new[] {typeof(TService1), typeof(TService2)}, typeof(TConcreteType), lifeStyle);
+      }
+
+      public void Register<TService1, TService2, TService3, TConcreteType>(LifeStyle lifeStyle)
+      {
+         Register(new[] {typeof(TService1), typeof(TService2), typeof(TService3)}, typeof(TConcreteType), lifeStyle);
+      }
+
+      public void Register<TService1, TService2, TService3, TService4, TConcreteType>(LifeStyle lifeStyle)
+      {
+         Register(new[] {typeof(TService1), typeof(TService2), typeof(TService3), typeof(TService4)}, typeof(TConcreteType), lifeStyle);
+      }
+
+      public void RegisterImplementationOf<TInterface>(TInterface component) where TInterface : class
+      {
+         AutofacBuilder.RegisterInstance(component).ExternallyOwned();
+      }
+
+      public void Register(IReadOnlyCollection<Type> serviceTypes, Type concreteType, LifeStyle lifeStyle)
+      {
+         register(serviceTypes, concreteType, lifeStyle);
+      }
+
+      public void RegisterImplementationOf<TInterface>(TInterface component, string key) where TInterface : class
+      {
+         AutofacBuilder.RegisterInstance(component).Keyed<TInterface>(key).ExternallyOwned();
+      }
+
+      public void RegisterOpenGeneric(Type serviceType, Type concreteType)
+      {
+         AutofacBuilder.RegisterGeneric(concreteType).As(serviceType).ExternallyOwned();
+      }
+
+      private void register(IReadOnlyCollection<Type> serviceTypes, Type concreteType, LifeStyle lifeStyle, string key = null)
+      {
+         var builder = AutofacBuilder
+            .RegisterType(concreteType)
+            .As(serviceTypes.ToArray())
+            .ExternallyOwned();
+
+         if (!string.IsNullOrEmpty(key))
+            builder.Keyed(key, concreteType);
+
+         if (lifeStyle == LifeStyle.Singleton)
+            builder.SingleInstance();
+
+         if(_hasActivationHooks)
+            _activationHooks.Each(x=> builder.OnActivated(x.OnActivated));
```

Comment:
I have defined activation hooks that allow me to register object when the are created. Very easy